subsys/bluetooth/host/hci_core.c

```diff
@@ -401,6 +405,72 @@ const bt_addr_le_t *bt_lookup_id_addr(u8_t id, const bt_addr_le_t *addr)
 }
 #endif /* CONFIG_BT_OBSERVER || CONFIG_BT_CONN */
 
+#if defined(CONFIG_BT_ADV_EXT)
+u8_t bt_le_ext_adv_get_index(struct bt_le_ext_adv *adv)
+{
+	u8_t index = adv - adv_pool;
+
+	__ASSERT(index < ARRAY_SIZE(adv_pool), "Invalid bt_adv pointer");
+	return index;
+}
+
+static struct bt_le_ext_adv *adv_new(void)
+{
+	struct bt_le_ext_adv *adv = NULL;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(adv_pool); i++) {
+		if (atomic_test_bit(adv_pool[i].flags, BT_ADV_CREATED)) {
```

Comment:
Shouldn't this be `if (!atomic_test_bit(...)`? This leads me to wonder if you tested this, since the way it's now it seems you'd never be able to allocate a new context.

Actually, to make this pre-emptible thread safe wouldn't `if (!atomic_test_and_set_bit(...)` be the best option? In that case you'll need to fix the memset that comes later though.

Reply:
Using the atomic_test_and_set_bit makes sense, but it becomes trickier to do the init when memset cannot be used. A will leave it using !atomic_test_bit(..). We should at some point address pre-emptible thread safety, and then fix this point, as well as the matching for connection objects.